src/api/IInput.js

```diff
@@ -27,6 +27,24 @@
         return true;
     };
 
+    IInput.prototype.hasValue = function () {
+        switch (this.type()) {
```

Comment:
this.type may not exist...

You want to provide a default implementation here and then override it as needed on the various implementations.

IOW Does Slider have a "type()"?